### PR TITLE
added PageFactory task solution

### PR DESCRIPTION
**Pull request by tatsianabeliai**

### CODE REVIEW

File: src/main/java/TestBase.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+import org.testng.annotations.AfterSuite;
+import org.testng.annotations.BeforeSuite;
+
+import java.util.concurrent.TimeUnit;
+
+public class TestBase {
+    public static final int DEFAULT_TIMEOUT = 10;
+    public WebDriver driver;
```
Comment (EugenBorisik) [inline src/main/java/TestBase.java]: Why these fields are public?

File: src/main/java/TestBase.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    protected String baseUrl;
+
+    public TestBase(String baseUrl) {
```
Comment (EugenBorisik) [inline src/main/java/TestBase.java]: Why is public?

File: src/main/java/TestBase.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        this.baseUrl = baseUrl;
+    }
+
+    public void waitForAllElements() {
```
Comment (EugenBorisik) [inline src/main/java/TestBase.java]: Do not understand, why do you need such waiter?
You set up page load time and implicit waiter for your WebDriver instance, so remove this waiter.

File: src/main/java/pages/EmailsFactoryPage.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package pages;
+
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+
+public class EmailsFactoryPage {
```
Comment (EugenBorisik) [inline src/main/java/pages/EmailsFactoryPage.java]: Do not use plurals in names

File: src/main/java/pages/LoginFactoryPage.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    public WebElement loginLink;
+
+    public LoginFactoryPage(WebDriver driver) {
+        PageFactory.initElements(driver, this);
+    }
+
+    public void login(String userName, String password, String domain) {
+        usernameField.sendKeys(userName);
+        passwordField.sendKeys(password);
+        Select loginDomainDrop = new Select(loginDomainDropDown);
```
Comment (EugenBorisik) [inline src/main/java/pages/LoginFactoryPage.java]: Create private method (selectDomain(String domain) f.e.) for lines 32-33

File: src/test/java/PageFactoryTest.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import pages.EmailsFactoryPage;
+import pages.LoginFactoryPage;
+
+public class PageFactoryTest extends TestBase {
+    public static String siteURL = "https://mail.ru/";
+    public String userName = "seleniumtests10";
+    public String password = "060788avavav";
+    public String domain = "mail.ru";
```
Comment (EugenBorisik) [inline src/test/java/PageFactoryTest.java]: Make private. Make as constants (8-11)

File: src/main/java/pages/LoginFactoryPage.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    public WebElement passwordField;
+
+    @FindBy(id = "mailbox__auth__button")
+    public WebElement loginButton;
+
+    @FindBy(id = "mailbox__login__domain")
+    public WebElement loginDomainDropDown;
+
+    @FindBy(id = "PH_authLink")
+    public WebElement loginLink;
```
Comment (EugenBorisik) [inline src/main/java/pages/LoginFactoryPage.java]: All WebElements should be private

File: src/test/java/PageFactoryTest.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        super(siteURL);
+    }
+
+    @Test
+    public void login() {
+        LoginFactoryPage loginFactoryPage = PageFactory.initElements(driver, LoginFactoryPage.class);
+        loginFactoryPage.login(userName, password, domain);
+        waitForAllElements();
+        EmailsFactoryPage emailsFactoryPage = PageFactory.initElements(driver, EmailsFactoryPage.class);
+        Assert.assertTrue(emailsFactoryPage.loginedUserNameAtTop.isDisplayed(), "Logout link is displayed so user is not logged in");
```
Comment (EugenBorisik) [inline src/test/java/PageFactoryTest.java]: All actions with WebElements should be encapsulated.

File: src/test/java/PageFactoryTest.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+    public PageFactoryTest() {
+        super(siteURL);
+    }
+
+    @Test
+    public void login() {
+        LoginFactoryPage loginFactoryPage = PageFactory.initElements(driver, LoginFactoryPage.class);
+        loginFactoryPage.login(userName, password, domain);
+        waitForAllElements();
```
Comment (EugenBorisik) [inline src/test/java/PageFactoryTest.java]: Remove this method, if you need explicit waiter, it should be placed in the end of login(...) method

File: src/test/java/PageFactoryTest.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        EmailsFactoryPage emailsFactoryPage = PageFactory.initElements(driver, EmailsFactoryPage.class);
+        Assert.assertTrue(emailsFactoryPage.loginedUserNameAtTop.isDisplayed(), "Logout link is displayed so user is not logged in");
+    }
+
+    @Test(dependsOnMethods = "login")
+    public void logout(){
+        EmailsFactoryPage emailsFactoryPage = PageFactory.initElements(driver, EmailsFactoryPage.class);
+        emailsFactoryPage.logout();
+        LoginFactoryPage loginFactoryPage = PageFactory.initElements(driver, LoginFactoryPage.class);
+        Assert.assertTrue(loginFactoryPage.loginLink.isDisplayed(), "No Login link so the user is logined");
```
Comment (EugenBorisik) [inline src/test/java/PageFactoryTest.java]: Should be encapsulated

**Review by EugenBorisik — CHANGES_REQUESTED** [review]

File: src/main/java/TestBase.java (hunk 1 above)
Comment (tatsianabeliai) [reply src/main/java/TestBase.java]: Made then protected

File: src/main/java/TestBase.java (hunk 3 above)
Comment (tatsianabeliai) [reply src/main/java/TestBase.java]: done

File: src/main/java/TestBase.java (hunk 2 above)
Comment (tatsianabeliai) [reply src/main/java/TestBase.java]: made protected

File: src/main/java/pages/EmailsFactoryPage.java (hunk 4 above)
Comment (tatsianabeliai) [reply src/main/java/pages/EmailsFactoryPage.java]: fixed

File: src/test/java/PageFactoryTest.java (hunk 8 above)
Comment (tatsianabeliai) [reply src/test/java/PageFactoryTest.java]: done

File: src/test/java/PageFactoryTest.java (hunk 10 above)
Comment (tatsianabeliai) [reply src/test/java/PageFactoryTest.java]: done

File: src/main/java/pages/LoginFactoryPage.java (hunk 7 above)
Comment (tatsianabeliai) [reply src/main/java/pages/LoginFactoryPage.java]: done

File: src/main/java/pages/LoginFactoryPage.java (hunk 5 above)
Comment (tatsianabeliai) [reply src/main/java/pages/LoginFactoryPage.java]: done

File: src/test/java/PageFactoryTest.java (hunk 6 above)
Comment (tatsianabeliai) [reply src/test/java/PageFactoryTest.java]: done

File: src/test/java/PageFactoryTest.java (hunk 9 above)
Comment (tatsianabeliai) [reply src/test/java/PageFactoryTest.java]: fixed

**Review by EugenBorisik — APPROVED** [review]

**Pull request merged**